File: iterator/csharp/IteratorCSharp/ConcreteIterator.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CSharpIteratorDotNetCore
{
    //Concrete Iterator implements traversal algorithm. Store traversal position
    class ConcreteIterator : Iterator
    {
        private ConcreteCollection _collection;
        private int _current = -1;

        public ConcreteIterator(ConcreteCollection collection)
        {
            this._collection = collection;
        }

        public override object CurrentItem()
        {
                return this._collection.GetItems()[_current];
        }

        public override int Key()
        {
            return this._current;
        }

        public override bool MoveNext()
        {
            if (++_current < _collection.GetItems().Count)
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        public override void Reset()
        {
            _current = -1;
        }
    }
}

File: iterator/java/JavaIterator/src/com/iterator/ConcreteIterator.java
package com.iterator;

import java.util.ArrayList;

public class ConcreteIterator implements Iterator {
    private int current = 0;
    private ConcreteCollection collection;

    public ConcreteIterator(ConcreteCollection collection)
    {
       this.collection = collection;
    }
    @Override
    public boolean hasNext() {
        return current < collection.getItems().size();
    }

    @Override
    public Integer getNext() {
        if(!hasNext())
        {
            return null;
        }
        Integer object = collection.getItems().get(current);
        current++;
        return object;
    }

}
